File: src-tauri/src/models/ffmpeg.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct JobProgress {
    pub percent: f64,
    pub frame: u64,
    pub fps: f64,
    pub bitrate: String,
    pub total_size: u64,
    pub current_time: f64,
    pub speed: String,
    pub eta: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct FFmpegCommand {
    pub input_path: String,
    pub output_path: String,
    pub video_codec: Option<String>,
    pub audio_codec: Option<String>,
    pub video_bitrate: Option<String>,
    pub audio_bitrate: Option<String>,
    pub resolution: Option<Resolution>,
    pub fps: Option<f64>,
    pub crf: Option<u32>,
    pub preset: Option<String>,
    pub hw_accel: Option<String>,
    pub filters: Vec<FilterSpec>,
    pub trim: Option<TrimSpec>,
    pub extra_args: Vec<String>,
    pub two_pass: bool,
    pub container: Option<String>,
    pub copy_video: bool,
    pub copy_audio: bool,
    pub no_video: bool,
    pub no_audio: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Resolution {
    pub width: u32,
    pub height: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub algorithm: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ai_model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ai_scale: Option<u32>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FilterSpec {
    pub id: String,
    pub name: String,
    pub category: Option<String>,   // "video" | "audio"
    pub params: HashMap<String, String>,
    pub enabled: bool,
    pub order: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TrimSpec {
    pub start: f64,
    pub end: f64,
    pub accurate: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct HWEncoder {
    pub name: String,
    pub codec: String,
    pub device: String,
    pub available: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct StreamExtraction {
    pub stream_index: u32,
    pub output_format: String,
    pub output_filename: Option<String>,
    pub options: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TrimSegment {
    pub start: f64,
    pub end: f64,
}

File: src/types/ffmpeg.ts
/** FFmpegコマンドの全パラメータ */
export interface FFmpegCommand {
  inputPath: string;
  outputPath: string;
  videoCodec?: string;
  audioCodec?: string;
  videoBitrate?: string;
  audioBitrate?: string;
  resolution?: Resolution;
  fps?: number;
  crf?: number;
  preset?: EncoderPreset;
  hwAccel?: HWAccelType;
  filters: FilterSpec[];
  trim?: TrimSpec;
  extraArgs: string[];
  twoPass: boolean;
  container?: string;
  copyVideo: boolean;
  copyAudio: boolean;
  noVideo: boolean;
  noAudio: boolean;
}

export interface Resolution {
  width: number;
  height: number;
  algorithm?: ScalingAlgorithm;
  aiModel?: AiUpscaleModel;
  aiScale?: 2 | 3 | 4;
}

export type ScalingAlgorithm = 'bilinear' | 'lanczos' | 'ai';

export type AiUpscaleModel =
  | 'realesr-animevideov3'
  | 'realesrgan-x4plus'
  | 'realesrgan-x4plus-anime';

export interface TrimSpec {
  start: number; // 秒
  end: number; // 秒
  accurate: boolean; // true=再エンコード, false=copy
}

/** フィルター定義 */
export interface FilterSpec {
  id: string; // UUID
  name: string; // FFmpegフィルター名
  displayName: string; // UI表示名（i18n キー）
  category: FilterCategory;
  params: Record<string, string>;
  enabled: boolean;
  order: number;
}

export type FilterCategory = 'video' | 'audio';

/** ビルトインフィルター定義（カタログ用） */
export interface FilterDefinition {
  name: string;
  displayNameKey: string;
  category: FilterCategory;
  description: string;
  params: FilterParamDefinition[];
  preview: boolean; // プレビュー対応か
}

export interface FilterParamDefinition {
  key: string;
  label: string; // i18n キー
  type: 'number' | 'string' | 'select' | 'boolean' | 'range' | 'file';
  default: string | number | boolean;
  min?: number;
  max?: number;
  step?: number;
  options?: { value: string; label: string }[];
  unit?: string; // "px", "dB", "%", "x"
}

export type EncoderPreset =
  | 'ultrafast'
  | 'superfast'
  | 'veryfast'
  | 'faster'
  | 'fast'
  | 'medium'
  | 'slow'
  | 'slower'
  | 'veryslow';

export type HWAccelType = 'nvenc' | 'qsv' | 'videotoolbox' | null;

/** ハードウェアエンコーダー情報 */
export interface HWEncoder {
  name: string;
  codec: string;
  device: string;
  available: boolean;
}

/** コーデック選択肢 */
export interface CodecOption {
  id: string;
  name: string; // "H.264", "H.265/HEVC"
  ffmpegName: string; // "libx264", "libx265"
  type: 'video' | 'audio';
  hwVariants?: string[]; // ["h264_nvenc", "h264_qsv"]
}

/** コンテナフォーマット選択肢 */
export interface ContainerOption {
  id: string;
  name: string; // "MP4"
  extension: string; // "mp4"
  supportedVideoCodecs: string[];
  supportedAudioCodecs: string[];
}
